controllers/nodehealthcheck_controller.go

```diff
@@ -318,47 +317,88 @@ func (r *NodeHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Requ
+	if err = r.deleteOrphanedRemediationCRs(nhc, append(notMatchingNodes, matchingNodes...), resourceManager, log); err != nil {
 		return result, err
 	}
 
 	// delete remediation CRs for healthy nodes
-	for _, node := range healthyNodes {
+	healthyCount := 0
+	for _, node := range notMatchingNodes {
 		log.Info("handling healthy node", "node", node.GetName())
 		remediationCRs, err := resourceManager.ListRemediationCRs(nhc, func(cr unstructured.Unstructured) bool {
-			return cr.GetName() == node.GetName()
+			return cr.GetName() == node.GetName() && resources.IsOwner(&cr, nhc)
 		})
 		if err != nil {
 			log.Error(err, "failed to get remediation CRs for healthy node", "node", node.Name)
 			return result, err
 		}
-		for _, remediationCR := range remediationCRs {
-			if err := r.deleteRemediationCR(&remediationCR, nhc, resourceManager, log); err != nil {
-				log.Error(err, "failed to delete remediation CR", "name", remediationCR.GetName())
-				return result, err
+
+		if len(remediationCRs) == 0 {
+			// when all CRs are gone, the node is considered healthy
+			resources.UpdateStatusNodeHealthy(node.GetName(), nhc, r.Recorder)
+			healthyCount++
+			continue
+		}
+
+		// set conditions healthy timestamp
+		conditionsHealthyTimestamp := resources.UpdateStatusNodeConditionsHealthy(&node, nhc, currentTime())
+		if conditionsHealthyTimestamp != nil {
+			// warn about pending CRs when all CRs have been deleted for some time already but still exist
+			doLog := true
+			logThreshold := conditionsHealthyTimestamp.Add(-logWhenCRPendingDeletionDuration)
+			for _, cr := range remediationCRs {
+				if cr.GetDeletionTimestamp() == nil || cr.GetDeletionTimestamp().After(logThreshold) {
+					doLog = false
+					// requeue when we need to log
+					var logIn time.Duration
+					if cr.GetDeletionTimestamp() != nil {
+						logIn = cr.GetDeletionTimestamp().Sub(logThreshold) + time.Second
+					} else {
+						logIn = logWhenCRPendingDeletionDuration + time.Second
+					}
+					finalRequeueAfter = utils.MinRequeueDuration(finalRequeueAfter, &logIn)
+				}
+			}
+			if doLog {
+				log.Info("Node conditions don't match unhealthy condition anymore, but node has remediation CR(s) with pending deletion, considering node as unhealthy")
 			}
 		}
+
+		if err = r.deleteRemediationCRs(remediationCRs, nhc, resourceManager, false, log); err != nil {
+			return result, err
+		}
 	}
 
+	nhc.Status.ObservedNodes = pointer.Int(len(selectedNodes))
+	nhc.Status.HealthyNodes = &healthyCount
+
 	// we are done in case we don't have unhealthy nodes
-	if len(unhealthyNodes) == 0 {
+	if len(matchingNodes) == 0 {
 		return result, nil
 	}
 
 	// check if we have enough healthy nodes
-	if minHealthy, err := intstr.GetScaledValueFromIntOrPercent(nhc.Spec.MinHealthy, len(nodes), true); err != nil {
+	skipRemediation := false
+	if minHealthy, err := intstr.GetScaledValueFromIntOrPercent(nhc.Spec.MinHealthy, len(selectedNodes), true); err != nil {
```

Comment:
I might mix up here, but isn't `len(selectedNodes)` is supposed to be equal to `healthyCount`?

Reply:
Or they might be different and it makes more sense to use `len(selectedNodes)` rather than `healthyCount`?

Reply:
Not sure if I understand your concern. `len(selectedNodes)` is the number of nodes this NHC is watching. We need it to calculate the absolute minimum number of healthy nodes in case it's set as percentage. Does that make sense?

Reply:
Yes, I have mixed up some stuff :)